keps/sig-node/4382-dra-numeric-model-simple-counter/README.md

```diff
@@ -0,0 +1,815 @@
+
+Adding another numeric model doesn't add any additional risks. There is a
+slight performance impact when checking whether the builtin controller of the
+numeric model needs to do something for a claim or NodeResourceCapacity object,
+but that impact should be minimal when the model is not in use.
+
+## Design Details
+
+Everything related to this numeric model is hosted in
+`k/k/staging/src/k8s.io/dynamic-resource-allocation`:
+
+- `numeric/counter`: the implementation of the builtin controller
+- `apis/counter`: the versioned type(s)
+- `apis/internal/counter`: the internal type for conversion
+
+The only modification elsewhere is registering the builtin controller in the
+dynamic resource scheduler plugin if (and only if) the model is enabled.
+
+### Capacity type
+
+```
+// Capacity defines how much of certain items are available respectively
+// required.
+type Capacity struct {
+	// ObjectMeta may be set when publishing capacity, but is not required.
+	// It is ignored when matching requests against available capacity.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// TypeMeta must have non-empty kind and apiVersion.
+	metav1.TypeMeta `json:",inline"`
+
+	// Count is the total number of available items per node when
+	// publishing capacity and the required number of items when requesting
+	// resources.
+	Count int64 `json:"count"`
+}
+```
+
+### Claim parameter type
+
+There is no separate type because the same `Count` field also makes sense when
+requesting items. The examples in the numeric parameters KEP show how this then
+gets used by a DRA driver.
+
+### Allocation result
+
+The following object gets serialized as JSON and stored in a claim's
+"status.allocation.resourcehandle.data" field:
+
+```
+// AllocationResult captures class and claim parameters and records which node
+// capacity instance was used. The JSON dump of this gets stored in the driver
+// handle field of an allocated claim.
+type AllocationResult struct {
+	// TypeMeta must have non-empty kind and apiVersion.
+	metav1.TypeMeta `json:",inline"`
+
+	// ClassParameters are the parameters from the time that the claim
+	// was allocated. They can be arbitrary setup parameters that are
+	// ignore by the counter controller.
+	ClassParameters runtime.RawExtension `json:"classParameters,omitempty"`
+
+	// ClaimParameters are the parameters from the time that the claim was
+	// allocated. Some of the fields were used by the counter controller to
+	// allocated resources, others can be arbitrary setup parameters.
+	ClaimParameters runtime.RawExtension `json:"claimParameters,omitempty"`
+
+	// NodeName is the name of the node providing the necessary resources.
+	NodeName string `json:"nodeName"`
+
+	// DriverName is the name of the driver on that node.
+	DriverName string `json:"driverName"`
+
+	// InstanceID is the unique ID chosen by that driver for the resource
+	// instance that was picked for the claim.
+	InstanceID types.UID `json:"instanceID"`
```

Comment:
To clarify: If a DRA plugin assigns multiple underlying devices (e.g. GPUs) to a claim (i.e. `Count > 1`) this does not map to a single resource. Does that mean that either:
1. The plugin needs to keep track of the association between `InstanceID` and the individual resources
2. This field needs to be repeatable?

If we are considering 1 above. Where is this state persisted so that this remains usable when things get deleted as below.

Reply:
It's responsibility of the DRA kubelet plugin to generated the instance ID and then map this field back to whatever the underlying hardware IDs are.

> this remains usable when things get deleted as below.

"as below" - which section are you referring to?

Reply:
I am refering to:
> This ensures that the claim
is usable also when class or claim parameters get deleted. It also avoids
having to grant the kubelet read permission for arbitrary parameter resources.

Reply:
That comment is mostly about `ClassParameters` and `ClaimParameters`: those are there for the DRA driver so that it can extract its configuration parameters.

The `InstanceID` needs to be checked by the DRA driver in case that it advertises more than one instance per node.

The numeric model implementation also uses it to match the allocated claims against the available resources in case that the kube-scheduler gets restarted and the numeric model needs to reconstruct which resources are in use.